cpp/src/io/utilities/data_sink.cpp

```diff
@@ -49,36 +58,50 @@ class file_sink : public data_sink {
 
   size_t bytes_written() override { return _bytes_written; }
 
-  [[nodiscard]] bool supports_device_write() const override { return _cufile_out != nullptr; }
+  [[nodiscard]] bool supports_device_write() const override
+  {
+    return _use_kvikio || _cufile_out != nullptr;
+  }
 
   [[nodiscard]] bool is_device_write_preferred(size_t size) const override
   {
-    return _cufile_out != nullptr && _cufile_out->is_cufile_io_preferred(size);
+    return _use_kvikio || (_cufile_out != nullptr && _cufile_out->is_cufile_io_preferred(size));
   }
 
-  void device_write(void const* gpu_data, size_t size, rmm::cuda_stream_view stream) override
+  std::future<void> device_write_async(void const* gpu_data,
+                                       size_t size,
+                                       rmm::cuda_stream_view stream) override
   {
     if (!supports_device_write()) CUDF_FAIL("Device writes are not supported for this file.");
 
-    _cufile_out->write(gpu_data, _bytes_written, size);
+    size_t offset = _bytes_written;
     _bytes_written += size;
+
+    if (_use_kvikio) {
+      // KvikIO's `pwrite()` returns a `std::future<size_t>` so we convert it
+      // to `std::future<void>`
+      return std::async(std::launch::deferred, [this, gpu_data, size, offset] {
+        _kvikio_file.pwrite(gpu_data, size, offset).get();
+      });
+    }
+    return _cufile_out->write_async(gpu_data, offset, size);
   }
 
-  std::future<void> device_write_async(void const* gpu_data,
-                                       size_t size,
-                                       rmm::cuda_stream_view stream) override
+  void device_write(void const* gpu_data, size_t size, rmm::cuda_stream_view stream) override
   {
     if (!supports_device_write()) CUDF_FAIL("Device writes are not supported for this file.");
 
-    auto result = _cufile_out->write_async(gpu_data, _bytes_written, size);
+    if (_use_kvikio) { return device_write_async(gpu_data, _bytes_written, stream).get(); }
+    _cufile_out->write(gpu_data, _bytes_written, size);
     _bytes_written += size;
-    return result;
   }
 
  private:
   std::ofstream _output_stream;
   size_t _bytes_written = 0;
   std::unique_ptr<detail::cufile_output_impl> _cufile_out;
+  kvikio::FileHandle _kvikio_file;
```

Comment:
Is this a kind of pointer? If so, we don't need to have the second bool var (`_use_kvikio`). Just initialize `_kvikio_file` to `nullptr` if it is not used.

Reply:
If it is not a pointer, just store it in a `shared_ptr` variable and initialize that variable as `nullptr`. After that, just check `if(_kvikio_file == nullptr)` instead of `if(_use_kvikio)`.

Reply:
Good point but instead of introducing a `unique_ptr`, we can use `_kvikio_file.closed()` to check if `_kvikio_file` has been initialized: https://github.com/rapidsai/cudf/pull/10593/commits/7e79d2c8d543ad91800e82662bd680aab6c1f985